### PR TITLE
utilizing `ruff` for formatting & linting – #20

### PR DESCRIPTION
**Pull request by jGaboardi**
utilizing `ruff` for formatting & linting – #20

* [x] run formatting & make initial PR
* [x] prelim review – *actually is only a single change in `base.py`*
* [x] run linting & re-push
* [ ] ...

### CODE REVIEW

**Comment by jGaboardi** [comment]
Running `ruff check fastpair` with the following rules ([standard](https://github.com/pysal/spopt/blob/fa782a5cb0ead740f3776c2f305522014515f5eb/pyproject.toml#L84) in the PySAL ecosystem):

```
lint.select = ["E", "F", "W", "I", "UP", "N", "B", "A", "C4", "SIM", "ARG"]
```

results in:

```
fastpair/__init__.py:2:1: UP009 [*] UTF-8 encoding declaration is unnecessary
fastpair/__init__.py:16:19: F401 `.base.FastPair` imported but unused; consider removing, adding to `__all__`, or using a redundant alias
fastpair/base.py:2:1: UP009 [*] UTF-8 encoding declaration is unnecessary
fastpair/base.py:36:1: UP010 [*] Unnecessary `__future__` imports `absolute_import`, `division`, `print_function` for target Python version
fastpair/base.py:36:1: I001 [*] Import block is un-sorted or un-formatted
fastpair/base.py:38:22: F401 [*] `operator.itemgetter` imported but unused
fastpair/base.py:46:7: N801 Class name `attrdict` should use CapWords convention
fastpair/base.py:50:14: UP008 Use `super()` instead of `super(__class__, self)`
fastpair/base.py:54:16: UP004 [*] Class `FastPair` inherits from `object`
fastpair/base.py:77:23: C408 Unnecessary `list` call (rewrite as a literal)
fastpair/base.py:101:21: F841 Local variable `res` is assigned to but never used
fastpair/base.py:119:16: E713 [*] Test for membership should be `not in`
fastpair/base.py:120:28: UP032 [*] Use f-string instead of `format` call
fastpair/base.py:124:16: E713 [*] Test for membership should be `not in`
fastpair/base.py:125:28: UP032 [*] Use f-string instead of `format` call
fastpair/test/test_fastpair.py:2:1: UP009 [*] UTF-8 encoding declaration is unnecessary
fastpair/test/test_fastpair.py:13:1: UP010 [*] Unnecessary `__future__` imports `absolute_import`, `division`, `print_function` for target Python version
fastpair/test/test_fastpair.py:13:1: I001 [*] Import block is un-sorted or un-formatted
fastpair/test/test_fastpair.py:17:44: F401 [*] `itertools.groupby` imported but unused
fastpair/test/test_fastpair.py:23:32: F401 [*] `scipy.unique` imported but unused
fastpair/test/test_fastpair.py:66:21: B905 [*] `zip()` without an explicit `strict=` parameter
fastpair/test/test_fastpair.py:82:5: N802 Function name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:99:26: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:106:24: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:117:24: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:130:24: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:137:29: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:144:42: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:153:38: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:171:44: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:179:9: E741 Ambiguous variable name: `l`
fastpair/test/test_fastpair.py:187:28: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:212:33: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:223:9: E741 Ambiguous variable name: `l`
fastpair/test/test_fastpair.py:224:9: F841 Local variable `res` is assigned to but never used
fastpair/test/test_fastpair.py:225:9: F841 Local variable `neigh` is assigned to but never used
fastpair/test/test_fastpair.py:270:25: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:280:13: F841 Local variable `myitem` is assigned to but never used
fastpair/test/test_fastpair.py:290:45: N803 Argument name `PointSet` should be lowercase
fastpair/test/test_fastpair.py:298:9: F841 Local variable `res` is assigned to but never used
```

There are actually far fewer linting issues than I anticipated and - from what I can tell - all are trivial fixes. Great job at writing future-proof code @carsonfarmer !

File: fastpair/test/test_fastpair.py [hunk 1]
```diff
@@ -220,9 +216,9 @@ def test_update_point(self, PointSet):
         assert old not in fp
         assert new in fp
         assert len(fp) == len(ps)  # Size shouldn't change
-        l = [(fp.dist(a, b), b) for a, b in zip(cycle([new]), ps)]
-        res = min(l, key=itemgetter(0))
-        neigh = fp.neighbors[new]
+        l_ = [(fp.dist(a, b), b) for a, b in zip(cycle([new]), ps)]
+        res = min(l_, key=itemgetter(0))  # noqa: F841
```
Comment (jGaboardi) [inline fastpair/test/test_fastpair.py:220]: I wasn't sure if `res` and `neigh` were actually needed for something, but not being testing so I left in and added logic for `ruff` to ignore.

File: fastpair/test/test_fastpair.py [hunk 2]
```diff
@@ -168,24 +164,24 @@ def test_all_closest_pairs(self, PointSet):
...
+        ps = point_set
         fp = FastPair().build(ps)
         rando = rand_tuple(len(ps[0]))
         neigh = fp._find_neighbor(rando)  # Abusing find_neighbor!
         dist = fp.dist(rando, neigh["neigh"])
         assert abs(dist - neigh["dist"]) < 1e-8
         assert len(fp) == len(ps)  # Make sure we didn't add a point...
-        l = [(fp.dist(a, b), b) for a, b in zip(cycle([rando]), ps)]
-        res = min(l, key=itemgetter(0))
+        l_ = [(fp.dist(a, b), b) for a, b in zip(cycle([rando]), ps)]
```
Comment (carsonfarmer) [inline fastpair/test/test_fastpair.py:175]: why the underscore here?

File: fastpair/test/test_fastpair.py (hunk 1 above)
Comment (carsonfarmer) [reply fastpair/test/test_fastpair.py:220]: they _were_ used in the following asserts, but have been commented out for reasons I am no longer aware of :P

Comment (carsonfarmer) [reply fastpair/test/test_fastpair.py:220]: So we could either turn those back on, or just drop res and neigh I suspect.

File: fastpair/test/test_fastpair.py [hunk 3]
```diff
@@ -277,23 +273,23 @@ def test_iter(self, PointSet):
...
         try:
-            myitem = fp[(2, 3, 4)]
+            fp[(2, 3, 4)]
         except KeyError as err:
             print(err)
 
         fp[ps[0]] = fp[ps[0]].neigh
         try:
-            fp[(2, 3, 4)] = fp[ps[0]].neigh
+            fp[ps[0]].neigh  # noqa: B018
```
Comment (carsonfarmer) [inline fastpair/test/test_fastpair.py:282]: this is here just to ensure it doesn't throw. I'm sure there are better/more idiomatic ways to do this

**Review by carsonfarmer — APPROVED** [review]

File: fastpair/test/test_fastpair.py (hunk 2 above)
Comment (jGaboardi) [reply fastpair/test/test_fastpair.py:175]: This is due to:

```
fastpair/test/test_fastpair.py:179:9: E741 Ambiguous variable name: `l`
```

If you prefer the `l` variable name, we can keep that and I'll throw a `# noqa: E741` in there to stop `ruff` from hollering.

File: fastpair/test/test_fastpair.py (hunk 3 above)
Comment (jGaboardi) [reply fastpair/test/test_fastpair.py:282]: > So we could either turn those back on, or just drop res and neigh I suspect.

I'll try to turn those back on and ....

> this is here just to ensure it doesn't throw. I'm sure there are better/more idiomatic ways to do this

maybe add some kind of `assert fp[ps[0]].neigh` for the something or even simply `assert hasttr(fp[ps[0]], "neigh")`

File: fastpair/test/test_fastpair.py (hunk 2 above)
Comment (carsonfarmer) [reply fastpair/test/test_fastpair.py:175]: naw it doesn't matter

File: fastpair/test/test_fastpair.py (hunk 3 above)
Comment (carsonfarmer) [inline fastpair/test/test_fastpair.py:282]:
```suggestion
            assert fp[ps[0]].neigh
```

**Review by carsonfarmer — APPROVED** [review]

**Pull request merged**